src/main/resources/data/computercraft/lua/rom/programs/poster.lua

```diff
@@ -0,0 +1,89 @@
+local printer = peripheral.find("poster_printer")
+
+local filename = (...)
+if not filename then
+    local programName = arg[0] or fs.getName(shell.getRunningProgram())
+    print("Usage: " .. programName .. " <filename>")
+    return
+end
+
+local image
+if not fs.exists(filename) then
+    if fs.exists(filename .. ".lit") then
+        filename = filename .. ".lit"
+    elseif fs.exists(filename .. ".nfp") then
+        filename = filename .. ".nfp"
+    else
+        print("File not found")
+        return
+    end
+end
+
+printer.reset()
+
+local shouldRemap = false
+if filename:sub(-4) == ".lit" then
+    local pictureTable = dofile(filename)
+    image = {}
+    for y = 1, pictureTable.height do
+        image[y] = {}
+        for x = 1, pictureTable.width do
+            image[y][x] = pictureTable[(y-1) * pictureTable.width + x]
+        end
+    end
+
+    for i = 1, #pictureTable.palette do
+        local color = pictureTable.palette[i]
+        local r = math.floor(color / 256 / 256)
+        local g = math.floor(color / 256) % 256
+        local b = color % 256
+        printer.setPaletteColor(i, r, g, b)
+    end
+
+    if pictureTable.name then printer.setLabel(pictureTable.name) end
+    if pictureTable.tooltip then printer.setTooltip(pictureTable.tooltip) end
+else
+    image = paintutils.loadImage(filename)
+    shouldRemap = true
+end
+
+if not image then
+    print("Failed to load image")
+    return
+end
+
+-- colors to minecraft mapcolor id
+local colorMap = {
+    [colors.white] = 8,
+    [colors.orange] = 15,
+    [colors.magenta] = 16,
+    [colors.lightBlue] = 17,
+    [colors.yellow] = 18,
+    [colors.lime] = 19,
+    [colors.pink] = 20,
+    [colors.gray] = 21,
+    [colors.lightGray] = 22,
+    [colors.cyan] = 23,
+    [colors.purple] = 24,
+    [colors.blue] = 25,
+    [colors.brown] = 26,
+    [colors.green] = 27,
+    [colors.red] = 28,
+    [colors.black] = 29,
+}
+
+for y = 1, #image do
+    for x = 1, #image[y] do
+        local color = image[y][x]
+        if shouldRemap then
+            local mapColor = colorMap[color]
+            if mapColor then
+                printer.setPixel(x, y, mapColor)
```

Comment:
This applies to PosterPrinterPeripheral too - I think that calling one setPixel per pixel is not what we should be encouraging. Instead, consider something like providing bulk update methods:

- `setPixels(string)` - takes a 16384 byte string (don't error if it's less, just don't write anything past that point). Then images can be downloaded in binary mode and directly copied.
- `setLine(y, string)` - takes a 128 byte string and assigns the line at that Y position to its contents. This is where it will matter if your posters are row-major or column-major, as ideally you want this to just be a copy. 

Of course for both of these you'd still need to scan the strings to validate them.

Another idea could be to provide a term-like API, but I think this is an unnecessary complexity that should be done in Lua instead, as having the term calls in Java would require them all being on the main thread, and thus slower.

Reply:
Wholly agree, I definitely want bulk update methods, I was just doing setPixel while we decided what we wanted for these.